daml-lf/engine/src/main/scala/com/digitalasset/daml/lf/engine/ConcurrentCompiledPackages.scala

```diff
@@ -55,7 +55,7 @@ final class ConcurrentCompiledPackages extends MutableCompiledPackages {
         val pkgId: PackageId = toCompile.head
         toCompile = toCompile.tail
 
-        if (!_packages.contains(pkgId)) {
+        if (!_packages.containsKey(pkgId)) {
```

Comment:
You might get better type safety if you convert this to a Scala map with `.asScala`.

Reply:
Right see the PR description, happy to make the change if someone tells me this is the right thing to do and has no performance impliciations.

Reply:
It's just a thin wrapper; shouldn't have any noticeable impact. The relevant implementation is in [`JMapWrapperLike`](https://github.com/scala/scala/blob/07260fe84557f316308086f2c95532bef173266a/src/library/scala/collection/convert/Wrappers.scala#L261-L294) and [`MapLike`](https://github.com/scala/scala/blob/07260fe84557f316308086f2c95532bef173266a/src/library/scala/collection/MapLike.scala#L153). `contains(x)` calls `get(x).isDefined`, and `get` just calls the underlying Map's `get` and wraps it in an `Option`.

Reply:
It is probably fine to define _package as follow:
``
  private[this] val _packages: collection.concurrent.Map[PackageId, Package] =
    new ConcurrentHashMap[PackageId, Package]().asScala
``

This uses a wrapper over the java collection, which does not change actual complexity. 
Anyway, the bug, you discovered, shows that performance is probably not critical here.

Reply:
Thanks! I’ve changed it. The Scala API only appears to have a `putIfAbsent` not a `computeIfAbsent` so I had to add an explicit check.